pkg/applier/kpt_applier.go

```diff
@@ -512,57 +514,60 @@ func (a *Applier) sync(ctx context.Context, objs []client.Object) (map[schema.Gr
 		}
 		gvks[resource.GetObjectKind().GroupVersionKind()] = struct{}{}
 	}
-	if a.errs == nil {
+
+	errs := a.Errors()
+	if errs == nil {
 		klog.V(4).Infof("all resources are up to date.")
 	}
-
 	if s.Empty() {
 		klog.V(4).Infof("The applier made no new progress")
 	} else {
 		klog.Infof("The applier made new progress: %s", s.String())
 		objStatusMap.Log(klog.V(0))
 	}
-	return gvks, a.errs
+	return gvks, errs
 }
 
+// Errors returns the errors encountered during the last apply or current apply
+// if still running.
 // Errors implements Interface.
-// Errors returns the errors encountered during apply.
 func (a *Applier) Errors() status.MultiError {
-	// TODO: Make read/write of a.errs thread-safe
-	// Return a copy
+	a.errorMux.RLock()
+	defer a.errorMux.RUnlock()
+
+	// Return a copy to avoid persisting caller modifications
 	return status.Append(nil, a.errs)
 }
 
-// Syncing implements Interface.
-// Syncing returns whether the applier is syncing.
-func (a *Applier) Syncing() bool {
-	return a.syncing
+func (a *Applier) addError(err error) {
+	a.errorMux.Lock()
+	defer a.errorMux.Unlock()
+
+	if _, ok := err.(status.Error); !ok {
+		// Wrap as an applier.Error to indidate the source of the error
+		err = Error(err)
+	}
+
+	a.errs = status.Append(a.errs, err)
+}
+
+func (a *Applier) invalidateErrors() {
+	a.errorMux.Lock()
+	defer a.errorMux.Unlock()
+
+	a.errs = nil
 }
 
 // Apply implements Interface.
 func (a *Applier) Apply(ctx context.Context, desiredResource []client.Object) (map[schema.GroupVersionKind]struct{}, status.MultiError) {
-	// Clear the `errs` field at the start.
-	a.errs = nil
-	// Set the `syncing` field to `true` at the start.
-	a.syncing = true
-
-	defer func() {
-		// Make sure to clear the `syncing` field before `Apply` returns.
-		a.syncing = false
-	}()
-
-	a.mux.Lock()
-	defer a.mux.Unlock()
-
-	// Pull the actual resources from the API server to compare against the
-	// declared resources. Note that we do not immediately return on error here
-	// because the Applier needs to try to do as much work as it can on each
-	// cycle. We collect and return all errors at the end. Some of those errors
-	// are transient and resolve in future cycles based on partial work completed
-	// in a previous cycle (eg ignore an error about a CR so that we can apply the
-	// CRD, then a future cycle is able to apply the CR).
-	// TODO: Here and elsewhere, pass the MultiError as a parameter.
-	return a.sync(ctx, desiredResource)
+	a.applyMux.Lock()
+	defer a.applyMux.Unlock()
+
+	// Ideally we want to avoid invalidating errors that will continue to happen,
+	// but for now, invalidate all errors until they recur.
+	// TODO: improve error cache invalidation to make rsync status more stable
+	a.invalidateErrors()
```

Comment:
What's the plan for how we want to handle this in the long run?

Reply:
I'm writing a doc. But the TLDR is to do more careful invalidation by caching what event type and object added/owns the error, and not invalidating that error until we see that event again, or the apply ends.